plugins/obs-ffmpeg/obs-ffmpeg-afilter.cpp

```diff
@@ -0,0 +1,963 @@
+class FilterGraph {
+	AVFilterGraph *graph_ = nullptr;
+	AVFilterContext *inputFilter_ = nullptr;
+	AVFilterContext *mainFilter_ = nullptr;
+	AVFilterContext *outputFilter_ = nullptr;
+
+	obs_audio_info oai_ = obs_audio_info{};
+
+public:
+	FilterGraph() {}
+	~FilterGraph();
+
+	bool initFilter(obs_data *settings, const obs_audio_info *oai);
+
+	obs_audio_data *
+	filterAudio(obs_audio_data *audio,
+		    array<vector<float>, MAX_AV_PLANES> &frame_datas, bool end);
+};
+
+class auto_adapt_string {
+	std::string buf_;
+
+public:
+	auto_adapt_string() {}
+
+	auto_adapt_string(const char *str)
+	{
+		if (str)
+			buf_ = str;
+	}
+
+	auto_adapt_string(const string &str) { buf_ = str; }
+
+	template<class T> auto_adapt_string(const T &val)
+	{
+		buf_ = to_string(val);
+	}
+
+	operator const string &() const { return buf_; }
+	operator const char *() const { return buf_.c_str(); };
+};
+
+int64_t speaker_layout_to_av_ch_layout(speaker_layout x)
+{
+	switch (x) {
+	case speaker_layout::SPEAKERS_MONO:
+		return AV_CH_LAYOUT_MONO;
+	case speaker_layout::SPEAKERS_UNKNOWN:
+	case speaker_layout::SPEAKERS_STEREO:
+		return AV_CH_LAYOUT_STEREO;
+	case speaker_layout::SPEAKERS_2POINT1:
+		return AV_CH_LAYOUT_2POINT1;
+	case speaker_layout::SPEAKERS_4POINT0:
+		return AV_CH_LAYOUT_4POINT0;
+	case speaker_layout::SPEAKERS_4POINT1:
+		return AV_CH_LAYOUT_4POINT1;
+	case speaker_layout::SPEAKERS_5POINT1:
+		return AV_CH_LAYOUT_5POINT1;
+	case speaker_layout::SPEAKERS_7POINT1:
+		return AV_CH_LAYOUT_7POINT1;
+	default:
+		return speaker_layout_to_av_ch_layout(
+			speaker_layout::SPEAKERS_UNKNOWN);
+	}
+}
+
+speaker_layout av_ch_layout_to_speaker_layout(int x)
+{
+	switch (x) {
+	case AV_CH_LAYOUT_MONO:
+		return speaker_layout::SPEAKERS_MONO;
+	case AV_CH_LAYOUT_STEREO:
+		return speaker_layout::SPEAKERS_STEREO;
+	case AV_CH_LAYOUT_2POINT1:
+		return speaker_layout::SPEAKERS_2POINT1;
+	case AV_CH_LAYOUT_4POINT0:
+		return speaker_layout::SPEAKERS_4POINT0;
+	case AV_CH_LAYOUT_4POINT1:
+		return speaker_layout::SPEAKERS_4POINT1;
+	case AV_CH_LAYOUT_5POINT1:
+		return speaker_layout::SPEAKERS_5POINT1;
+	case AV_CH_LAYOUT_7POINT1:
+		return speaker_layout::SPEAKERS_7POINT1;
+	default:
+		return speaker_layout::SPEAKERS_UNKNOWN;
+	}
+}
+
+class FFAFilter : public IAudioFilter<FFAFilter> {
+	static bool obs_properties_add_avfilter(obs_properties *props,
+						obs_data *settings,
+						const obs_audio_info &oai,
+						const AVFilter *filter);
+
+	static bool obs_properties_hide_other_avclass(obs_properties *props,
+						      const AVFilter *filter);
+
+	mutex graph_mutex_;
+	unique_ptr<FilterGraph> filter_graph_;
+	unique_ptr<FilterGraph> next_filter_graph_;
+	string current_filter_name_;
+	array<vector<float>, MAX_AV_PLANES> last_audio_data_bufs_;
+	obs_audio_data last_audio_data_;
+
+public:
+	FFAFilter(obs_data_t *settings, obs_source_t *filter);
+	~FFAFilter() {}
+
+	static const char *Name(void *type_data)
+	{
+		return "FFMPEG AudioFilter";
```

Comment:
I'd vote this pull from locale (so the name can be translated), and should be proper: "FFmpeg Audio Filter"